src/mmap.cc

```diff
@@ -31,21 +34,71 @@ static void DontFree(char* data, void* hint) {
 NAN_METHOD(Alloc) {
   Nan::HandleScope scope;
 
-  int len = info[0]->Uint32Value();
+  std::size_t len;
+
+  if (node::Buffer::HasInstance(info[0])) {
+    len = *reinterpret_cast<std::size_t *>(node::Buffer::Data(info[0]));
+  } else {
+    len = info[0]->Uint32Value();
+  }
+
   int prot = info[1]->Uint32Value();
   int flags = info[2]->Uint32Value();
   int fd = info[3]->Int32Value();
-  int off = info[4]->Uint32Value();
+
+  std::int64_t off;
+
+  if (node::Buffer::HasInstance(info[4])) {
+    off = *reinterpret_cast<std::int64_t *>(node::Buffer::Data(info[4]));
+  } else {
+    off = info[4]->Int32Value();
+  }
 
   void* res = mmap(NULL, len, prot, flags, fd, off);
   if (res == MAP_FAILED)
     return Nan::ThrowError("mmap() call failed");
 
-  info.GetReturnValue().Set(Nan::NewBuffer(
-        reinterpret_cast<char*>(res),
-        len,
+  const std::size_t max_len = Nan::imp::kMaxLength;
+  Local<Value> ret;
+
+  if (len <= max_len) {
+    ret = Nan::NewBuffer(
+      reinterpret_cast<char*>(res),
+      len,
+      FreeCallback,
+      reinterpret_cast<void*>(static_cast<intptr_t>(len))).ToLocalChecked();
+  } else {
+    std::size_t ret_arr_len;
+
+    if ((len % max_len) == 0) {
+      ret_arr_len = len / max_len;
+    } else {
+      ret_arr_len = (len / max_len) + 1u;
+    }
+
+    Local<Array> ret_arr = Nan::New<Array>(ret_arr_len);
+
+    for (std::size_t i = 0u; i < ret_arr_len; i++) {
+      std::size_t curr_len;
+
+      if (i == (ret_arr_len - 1) &&
```

Comment:
I'd rather decrement `len` on each iteration and use `min(len, max_len)` for the buffer size.